config.yml

```diff
@@ -1,30 +1,38 @@
----
-  synapse:
-    master_fileview: 'syn20446927'
-    manifest_filename: 'manifests/synapse_storage_manifest.csv'
-    api_creds: 'syn21088684'
+synapse:
+  master_fileview: 'syn20446927'
+  manifest_filename: 'manifests/synapse_storage_manifest.csv'
+  api_creds: 'syn21088684'
 
-  model:
-    input:
-      location: 'schema_org_schemas/HTAN.jsonld'
-      file_type: 'local'
-      log_location: 'json_schema_logs/json_schema_log.json'
-    biothings:
-      location: 'schema_org_schemas/biothings.jsonld'
-    demo:
-      location: 'schema_org_schemas/HTAPP.jsonld'
-      file_type: 'local'
-      validation_file_location: 'validation_schemas/minimalHTAPPJSONSchema.json'
-      valid_manifest: 'manifests/HTAPP_manifest_valid.csv'
+model:
+  input:
+    location: 'schema_org_schemas/HTAN.jsonld'
+    file_type: 'local'
+    log_location: 'json_schema_logs/json_schema_log.json'
+  biothings:
+    location: 'schema_org_schemas/biothings.jsonld'
+  demo:
+    location: 'schema_org_schemas/HTAPP.jsonld'
+    file_type: 'local'
+    validation_file_location: 'validation_schemas/minimalHTAPPJSONSchema.json'
+    valid_manifest: 'manifests/HTAPP_manifest_valid.csv'
 
-  style:
-    google_manifest:
-      req_bg_color:
-        red: 0.9215
-        green: 0.9725
-        blue: 0.9803
-      opt_bg_color:
-        red: 1.0
-        green: 1.0
-        blue: 0.9019
-      master_template_id: '1LYS5qE4nV9jzcYw5sXwCza25slDfRA1CIg3cs-hCdpU'
+style:
+  google_manifest:
+    req_bg_color:
+      red: 0.9215
+      green: 0.9725
+      blue: 0.9803
+    opt_bg_color:
+      red: 1.0
+      green: 1.0
+      blue: 0.9019
+    master_template_id: '1LYS5qE4nV9jzcYw5sXwCza25slDfRA1CIg3cs-hCdpU'
+
+# Do not change the 'definitions' section unless you know what you're doing
+definitions:
+  root_dir: "./"
+  synapse_config: ".synapseConfig"
+  creds_path: "credentials.json"
+  token_pickle: "token.pickle"
+  data_path: "data/"
+  service_acct_creds: "quickstart-1560359685924-198a7114b6b5.json"
```

Comment:
Trivial note:

Do you think it would make more sense to move this "definitions" section to the top of the YAML file? I'm suggesting this because all paths below it (for ex. `[model][input][location]`) is/will be relative to the `root_dir` / `data_path`.

Reply:
Because we've moved away from everything being in the same folder relative to the repo, maybe we can drop the concept of a root folder. All paths in the config should be absolute or relative to the config file. What do you think?

Reply:
Here's a quick search for `ROOT_DIR` in the `schematic/` subfolder. If we exclude the matches in `configuration.py` or Markdown files, there aren't many other references to `ROOT_DIR`. It would be easy to remove these. 

```
9 results - 5 files

schematic/configuration.py:
    8          # this is the project's root directory
    9:         self.ROOT_DIR = None
   10          # path to config.yml file

   92          # Update module-level configuration constants
   93:         self.ROOT_DIR = self.DATA["definitions"]["root_dir"]
   94          self.CONFIG_PATH = config_path

  100          # Normalize all file paths as absolute file paths
  101:         self.ROOT_DIR = self.normalize_path(self.ROOT_DIR)
  102          self.CONFIG_PATH = self.normalize_path(self.CONFIG_PATH)

schematic/models/README.md:
  12  
  13: _Note: `CONFIG.CONFIG_PATH`, `CONFIG.ROOT_DIR`, `CONFIG.DATA_PATH` (constants) are loaded from the configuration YAML file._
  14  

  32      syn.login()
  33:     syn.get(SYN_CREDS, downloadLocation = CONFIG.ROOT_DIR)
  34      print("Stored Google API credentials.")

schematic/models/examples/metadata_usage.py:
  37      syn.login()
  38:     syn.get(API_CREDS, downloadLocation = CONFIG.ROOT_DIR)
  39      print("Stored Google API credentials.")

schematic/synapse/examples/store_usage.py:
  8  
  9: PATH_TO_SYN_CONF = os.path.join(CONFIG.ROOT_DIR, CONFIG.SYNAPSE_CONFIG_PATH)
  10  

schematic/utils/google_api_utils.py:
  70          syn.login()
  71:         syn.get(SYN_CREDS, downloadLocation = CONFIG.ROOT_DIR)
  72          print("Downloaded Google API credentials file.")
```

Reply:
That makes sense @BrunoGrandePhD!

So we'll get rid of the `root_dir`attribute from `definitions`? But we still want to allow users to configure other parameters in definitions, like the `data_path` and stuff right?

Reply:
> So we'll get rid of the `root_dir` attribute from `definitions`? 

Exactly, that's what I had in mind. Otherwise, it'll be confusing for users to know if non-absolute paths are relative to the config file or the specified root directory. In the code, I convert all file paths under `definitions` to absolute ones relative to the config file (see [code](https://github.com/BrunoGrandePhD/schematic/blob/add_csv_to_schemaorg_cli/schematic/configuration.py#L58-L65
)). I should consider extending this to all file paths that appear in the config file, not just the ones under `definitions`. Otherwise, we still face inconsistent behavior. 

On this note, I should look into all instances of `os.path.join()` in `schematic` to make sure that the paths are all being updated consistently. 

> But we still want to allow users to configure other parameters in definitions, like the `data_path` and stuff right?

👍

Reply:
That sounds great. Yes, it's important to ensure consistency (with respect to file paths) across all the modules in the app. Feel free to open another PR for that if you'd like. Else, new commits on this should work just as well. Thank you!

Reply:
Actually, I want to go back on what I said about `data_path`. I noticed that most of the `os.path.join()` calls are concatenating `data_path` with another attribute from the config (_e.g._ `CONFIG["model"]["input"]["location"]`). In fact, that's all `data_path` is ever used for: for being concatenated with another attribute in the config (see search results below), except for `csv_to_schemaorg_usage.py`, which needs to be updated to use a CLI. 

I propose that we remove `data_path` from the config, and replace all instances of `os.path.join(CONFIG.DATA_PATH, CONFIG["X"]["Y"]["Z"])` with just `CONFIG["X"]["Y"]["Z"]`.

```
schematic/manifest/examples/generator_usage.py:
  10: PATH_TO_JSONLD = os.path.join(CONFIG.DATA_PATH, CONFIG["model"]["input"]["location"])

schematic/models/examples/metadata_usage.py:
  13: MM_LOC = os.path.join(CONFIG.DATA_PATH, CONFIG["model"]["input"]["location"])
  54: HTAPP_VALIDATION_SCHEMA = os.path.join(CONFIG.DATA_PATH, CONFIG["model"]["demo"]
  59: HTAPP_SCHEMA = os.path.join(CONFIG.DATA_PATH, CONFIG["model"]["demo"]["location"])
  72: MANIFEST_PATH = os.path.join(CONFIG.DATA_PATH, CONFIG["model"]["demo"]["valid_manifest"])

schematic/schemas/generator.py:
  652:         json_schema_log_file = os.path.join(CONFIG.DATA_PATH, '', CONFIG["model"]["input"]["log_location"])

schematic/schemas/examples/explorer_usage.py:
  10: PATH_TO_JSONLD = os.path.join(CONFIG.DATA_PATH, CONFIG["model"]["input"]["location"])
  47: gv_digraph.render(os.path.join(CONFIG.DATA_PATH, '', 'viz/example-GV'), view=True)

schematic/schemas/examples/generator_usage.py:
  7: PATH_TO_JSONLD = os.path.join(CONFIG.DATA_PATH, CONFIG["model"]["input"]["location"])

schematic/synapse/store.py:
  80:             self.manifest = os.path.join(CONFIG.DATA_PATH, CONFIG["synapse"]["manifest_filename"])

schematic/synapse/examples/store_usage.py:
  46: MANIFEST_LOC = os.path.join(CONFIG.DATA_PATH, '', CONFIG["synapse"]["manifest_filename"])

schematic/utils/examples/csv_to_schemaorg_usage.py:
  13: base_schema_path = os.path.join(CONFIG.DATA_PATH, '', CONFIG["model"]["biothings"]["location"])
  19  schema_extensions_csv = [
  20:                         os.path.join(CONFIG.DATA_PATH, '', 'csv/example.csv')
  21                          ]
```

Reply:
Right, that makes sense. Initially, we were allowing the "folder" to be configured in `definitions.py` and the "relative file path" (relative to the directory specified in `definitions`) to be configured in the `config.yml`.

But you're right, what you're suggesting makes complete sense. We can remove the `data_path` parameter, and start specifying full paths (ex. `data/schema_org_schemas/example.jsonld`) and such) for other parameters in the `config.yml` file.

Reply:
I agree integrating paths in config.yml would make it more streamlined .

Looks like this PR is approved but there is still more work on it, so I assume it's not ready to be merged yet?

Reply:
Yeah, it's not ready yet. I need to wrap up the following:

- [ ] Replace remaining references to the `definitions.py` variables
- [ ] Remove `root_dir` and `data_path` references
- [ ] Track down all non-python files that need to be distributed with the package
  - [ ] Refer to these files using `pkg_resources` for portability
- [ ] Add CLI to remaining command-line tools (at least the most commonly used scripts)